### ISSUE TITLE
Impossible or Inaccurate Negative Values For Entity States

### ISSUE THREAD

**Issue by jazzyisj** [Pirate-Weather/pirateweather#394]
### Describe the bug

Sensors that should never have a negative value frequently report wildly inaccurate negative values.  I've seen this behaviour with humidly, barometric pressure, and cloud coverage sensors.  Maybe these are  points at which accurate data was probably not available?

<img width="1274" alt="image" src="https://github.com/user-attachments/assets/929b6391-4f8c-4eff-81d3-5fa9f3d91c8b" />

### Expected behavior

When sensor data is "unavailable" states should report that, not inaccurate numerical values.

### Actual behavior

Impossible negative values are reported.

### API Endpoint

Production

### Location

Windsor, Ontario, Canada

### Troubleshooting steps

- [X] I have searched this repository and [Home Assistant Repository](https://github.com/alexander0042/pirate-weather-ha) to see if the issue has already been reported.
- [X] I have read through the [API documentation](https://pirateweather.net/en/latest/API/) before opening this issue.
- [X] I have not included my API key anywhere in my issue.
- [X] I have written an informative title.

**Comment by cloneofghosts** [comment]
Thanks for opening this issue and sorry for the slow reply here. I know its been a few weeks since this issue was opened so am wondering if this issue is still occurring. If so does the issue happen on specific times or is it seemingly random? I'll ping @alexander0042 to investigate either way to look into this.

Taking a look at the code for cloud cover its supposed to be clipped between 0-1 so these large negative values shouldn't be occurring.
https://github.com/Pirate-Weather/pirate-weather-code/blob/e69a5a9219083beebaaf0e1f28a3b554dc405b02/API/responseLocal.py#L2621-L2634

**Comment by cloneofghosts** [comment]
> Yeah that is strange. I'm wondering if is happening on restarts?

That's a possibility but I'm not sure if the production endpoint still has a restart policy enabled on it? @alexander0042 would have to confirm if that's still the case.

I have a cloud cover sensor enabled on my HA setup and I haven't seen these issues with the large negative values though my refresh rate is 30m so maybe I'm skipping over the problem entirely.

**Comment by cloneofghosts** [comment]
Hi @jazzyisj were you able to test this over the weekend?

**Comment by jazzyisj** [comment]
Sorry I was not able to get to it on the weekend.  I just added the sensors I know had erroneous values back into my recorder.  Give me a few days to see if I can reproduce again.

**Comment by cloneofghosts** [comment]
While I'm running a version which uses the development endpoint instead of production I encountered the issue you saw over the weekend. I'm not exactly sure what causes the issue so will ping @alexander0042 to see if he can look into what's happening here.

**Comment by jazzyisj** [comment]
Whew, I'm not crazy!!  I've been checking every day and haven't been able to reproduce last few days.

Curious, did it happen on a system restart?

**Comment by cloneofghosts** [comment]
> Whew, I'm not crazy!! I've been checking every day and haven't been able to reproduce last few days.

I had it happen to me on the 18th/19th at around 12am EST. Seems to be a rare occurance as I hadn't happen in a while before then and haven't seen it after that.

> Curious, did it happen on a system restart?

You're talking about a API restart and not HA restart? I'm not sure if the API has automatic restarts setup anymore and the main developer will be having (or already has at this point) his second child so may be some time before he's able to reply.

**Comment by jazzyisj** [comment]
No, HA restart.  I thought maybe that's when it was happening to me but wasn't sure.

Congrats @alexander0042!  Yep it will be a while before we hear from him :-)

**Comment by cloneofghosts** [comment]
> No, HA restart. I thought maybe that's when it was happening to me but wasn't sure.

I don't have an automatic restart script setup on my HA install and I only usually update in the evenings so I don't think its caused by HA restarts. Its something on the APIs end but I'm not sure exactly what it might be.

> Congrats [@alexander0042](https://github.com/alexander0042)! Yep it will be a while before we hear from him :-)

I noticed the development endpoint is back up this morning so unless it fixed itself he might be around?

**Comment by cloneofghosts** [comment]
@alexander0042 mentioned yesterday that he's back working on this so will ping him again to see if he can figure out whatever the issue is.

**Comment by alexander0042** [comment]
I am back, and thanks for the kind words! Any more recurrences of this? That initial plot is super helpful, since I can compare it against my logs.

Re. the dev endpoint: while it'll always be less stable (by definition), I finally updated the way AWS handles the containers, and so it should be much more automatic now

**Comment by jazzyisj** [comment]
Sorry, I couldn't say at this point.  As I mentioned, I template all of my weather sensors  and have  already added filters to ignore any erroneous values myself after I noticed it happening a while back.  Once cloneofghosts confirmed he could duplicate it I turned off recording for my pirate weather sensors.  If you can't figure out what's going on with what you've got hit me up and I'll store the pirateweather data again for a while.

**Comment by cloneofghosts** [comment]
@alexander0042 I can post a screenshot of the cloud coverage sensor but looking through the sensor history I see that it happened last yesterday at midnight. Didn't happen today at midnight so might be a rare occurrence?

**Comment by alexander0042** [comment]
I've also added some sensors on my end, so see if I can track it down as well! Importantly, I significantly improved the monitoring today, which should really help to see if these occur in conjunction with anything on the back end.

**Comment by jazzyisj** [comment]
I just happened to be looking at my weather stuff when I caught a bad value for cloud coverage.

12:05 AM Windsor, Ontario

<img width="422" alt="Image" src="https://github.com/user-attachments/assets/d9adc388-0fdd-4a20-beb1-e1071ec51b43" />

**Comment by cloneofghosts** [comment]
@alexander0042 have you been able to figure anything out for this? I checked this morning and I see occurrences of this issue on the 14th and 16th. Both times it happened at 12:02am EDT so it seems to primarily occur around that time.

**Comment by trevorturk** [comment]
I'm also seeing a few sporadic cases where cloudCover is returning -999.0. Some example lat/lon and times, in case that's helpful. 

40.83,-81.54 at Mar 18, 2025 11:00:46 PM CDT
40.83,-81.54 at Mar 15, 2025 11:01:29 PM CDT
47.48,-92.46 at Mar 13, 2025 11:00:03 PM CDT

**Comment by cloneofghosts** [comment]
> I'm also seeing a few sporadic cases where cloudCover is returning -999.0. Some example lat/lon and times, in case that's helpful.

Will try pinging @alexander0042 to see if he has figured anything out. Something is causing certain variables to return NaN (the API replaces these with -999) and seems to happen around 12am EDT.

**Comment by cloneofghosts** [comment]
Just popping into say that I encountered this issue again this morning at 12:01am so the various changes that have been made to the API since March hasn't fixed this issue.

@alexander0042 have you been able to figure anything out for this?

**Comment by cloneofghosts** [comment]
I've been running my HA integration on the development endpoint for a while now and I haven't seen this issue in at least the last three weeks so maybe this is solved? Have either of you encountered this issue recently or can we close this issue?

**Comment by jazzyisj** [comment]
I'm on version 1.7.3
June 13 on a few different sensors (pressure, cloud, humidity, apparent temp)

<img width="296" alt="Image" src="https://github.com/user-attachments/assets/6ab546df-bbc4-45c7-9aa6-61d8d690f907" />

**Comment by alexander0042** [comment]
Oh shoot, I really thought that the V2.7 release would address this, since it improved a lot of stability things. Can you tell if that was a -999 value on the 13th, or some other number?

**Comment by jazzyisj** [comment]
This is humidity on June 13th.  The incorrect values appear at midnight and revert back to normal at 12:20 AM in all cases.

<img width="151" alt="Image" src="https://github.com/user-attachments/assets/19c33d33-bd1f-46e1-bd9a-5eb36bd62cf7" />

This is pressure.

<img width="197" alt="Image" src="https://github.com/user-attachments/assets/2a9e9113-d9d3-44b0-952d-6e80f313e79e" />

This is cloud coverage.

<img width="178" alt="Image" src="https://github.com/user-attachments/assets/e8914d60-3f8a-4f87-b809-6d3d7af3b1c7" />

This is apparent temperature.

<img width="215" alt="Image" src="https://github.com/user-attachments/assets/bb8a589e-e0a9-4ab5-9933-bec94fe45a6e" />

I looked at a handful of others (temperature, dew point, wind speed) and they appear to be correct values at the same time.

**Comment by trevorturk** [comment]
I saw a spike on June 7 and another smaller on on June 19, but both were actually minutely precipType!

**Comment by alexander0042** [comment]
@jazzyisj. that's really helpful here- the ones without issues are the hrrr subhourly values, which really helps with troubleshooting. I have a hunch for this, so let me try a couple things to see

**Comment by alexander0042** [comment]
Ok, isolated the issue here! I figured it related to the [indexing for the currently values](https://github.com/Pirate-Weather/pirate-weather-code/blob/47bb2329c4a45a814246f9006e2e79dc68887edf/API/responseLocal.py#L3624) , and this is somewhat true. What's happening is that when a request is made for midnight, the nearest index is also the first (0th) element in the interpolated array. Since currently is a linear interpolation between the two nearest hourly values, at midnight, this shows up as indexes -1 and 0, with a factor of 0 for the first one (aka, not used), and 1 for the second. I wasn't worried about this, since the -1 value has a factor of zero; however, since it sometimes draws a nan value, 0 * nan = nan, it propagates through as an error. 

The reason it's intermittent is that this only occurs when a request is made within ~2 minutes of midnight, so doesn't always appear. I'm going to put a fix in along with the time machine fix, and should push to prod later this week!
